src/lib/datocms/index.ts

```diff
@@ -72,35 +72,55 @@ type CollectionMeta = {
  * DatoCMS GraphQL API has a limit of 100 records per request.
  * This function uses pagination to get all records.
  * @see https://www.datocms.com/docs/content-delivery-api/pagination
+ * 
+ * @param {string} params.collection 
+ * - The name of the DatoCMS collection. For example, `"Pages"`
+ * @param {DocumentNode|string} params.fragment 
+ * - The GraphQL fragment to include for each record, For example `pageRouteFragment`.
  */
 export const datocmsCollection = async <CollectionType>({
   collection,
   fragment
 }: {
-  collection: string,
-  fragment: string
+  collection: string;
+  fragment: string | DocumentNode;
 }) => {
-  const { meta } = await datocmsRequest({
+  const {
+    meta,
+    records: [
+      { __typename: type } = { __typename: '' } // Collection might be empty
+    ]
+  } = await datocmsRequest<CollectionInfo>({
     query: parse(/* graphql */`
       query ${collection}Meta {
+        records: all${collection}(first: 1) { __typename }
         meta: _all${collection}Meta { count }
       }
    `)
-  }) as { meta: CollectionMeta };
-
+  });
   const recordsPerPage = 100; // DatoCMS GraphQL API has a limit of 100 records per request
   const totalPages = Math.ceil(meta.count / recordsPerPage);
   const records: CollectionType[] = [];
+  const fragmentDocument = typeof fragment === 'string'
+    ? parse(`fragment InlineFragment on ${type} { ${fragment} }`)
+    : fragment;
+  const { definitions } = fragmentDocument;
+  const fragmentDefinition = definitions
+    .find((definition): definition is FragmentDefinitionNode =>
+      definition.kind === Kind.FRAGMENT_DEFINITION
+    );
 
   for (let page = 0; page < totalPages; page++) {
     const data = await datocmsRequest({
       query: parse(/* graphql */`
+        ${print(fragmentDocument)}
```

Comment:
Insert fragment definition from `fragmentDocument`, which is either the `fragment` passed from an import from `@lib/datocms/types.ts` or the one created from a `string`;